pkg/ctl/cmdutils/nodegroup_filter.go

```diff
@@ -26,58 +25,81 @@ func NewNodeGroupFilter() *NodeGroupFilter {
 }
 
 // AppendGlobs appends globs for inclusion and exclusion rules
-func (f *NodeGroupFilter) AppendGlobs(includeGlobExprs, excludeGlobExprs []string, nodeGroups []*api.NodeGroup) error {
-	if err := f.AppendIncludeGlobs(nodeGroups, includeGlobExprs...); err != nil {
+func (f *NodeGroupFilter) AppendGlobs(includeGlobExprs, excludeGlobExprs, ngNames []string) error {
+	if err := f.AppendIncludeGlobs(ngNames, includeGlobExprs...); err != nil {
 		return err
 	}
 	return f.AppendExcludeGlobs(excludeGlobExprs...)
 }
 
 // AppendIncludeGlobs sets globs for inclusion rules
-func (f *NodeGroupFilter) AppendIncludeGlobs(nodeGroups []*api.NodeGroup, globExprs ...string) error {
-	return f.doAppendIncludeGlobs(f.collectNames(nodeGroups), "nodegroup", globExprs...)
+func (f *NodeGroupFilter) AppendIncludeGlobs(ngNames []string, globExprs ...string) error {
+	return f.doAppendIncludeGlobs(ngNames, "nodegroup", globExprs...)
+}
+
+// A stackLister lists nodegroup stacks
+type stackLister interface {
+	ListNodeGroupStacks() ([]manager.NodeGroupStack, error)
 }
 
 // SetExcludeExistingFilter uses stackManager to list existing nodegroup stacks and configures
 // the filter accordingly
-func (f *NodeGroupFilter) SetExcludeExistingFilter(stackManager *manager.StackCollection) error {
+func (f *NodeGroupFilter) SetExcludeExistingFilter(lister stackLister) error {
 	if f.ExcludeAll {
 		return nil
 	}
 
-	existing, err := stackManager.ListNodeGroupStacks()
+	existingStacks, err := lister.ListNodeGroupStacks()
 	if err != nil {
 		return err
 	}
 
-	return f.doSetExcludeExistingFilter(existing, "nodegroup")
+	var ngNames []string
+	for _, s := range existingStacks {
+		ngNames = append(ngNames, s.NodeGroupName)
+	}
+
+	return f.doSetExcludeExistingFilter(ngNames, "nodegroup")
 }
 
-// SetIncludeOrExcludeMissingFilter uses stackManager to list existing nodegroup stacks and configures
+// SetIncludeOrExcludeMissingFilter uses stackLister to list existing nodegroup stacks and configures
 // the filter to either explicitly exclude or include nodegroups that are missing from given nodeGroups
-func (f *NodeGroupFilter) SetIncludeOrExcludeMissingFilter(stackManager *manager.StackCollection, includeOnlyMissing bool, nodeGroups []eks.KubeNodeGroup) error {
-	existing, err := stackManager.ListNodeGroupStacks()
+func (f *NodeGroupFilter) SetIncludeOrExcludeMissingFilter(lister stackLister, includeOnlyMissing bool, clusterConfig *api.ClusterConfig) error {
+	stacks, err := lister.ListNodeGroupStacks()
 	if err != nil {
 		return err
 	}
 
-	remote := sets.NewString(existing...)
+	stackExists := func(name string) bool {
```

Comment:
Can we move this function out since SetIncludeOrExcludeMissingFilter is already a bit confusing...?

Reply:
Moved.